drivers/watchdog/wdt_wwdg_stm32.c

```diff
@@ -0,0 +1,272 @@
+ *  - counter = 0x40
+ *  - prescaler = 1
+ * The maximim timeout is calculated with:
+ *  - counter = 0x7F
+ *  - prescaler = 8
+ *
+ * E.g. for f_PCLK = 2MHz
+ *  t_WWDG_min = 1000 * ((0x40 & 0x3F) + 1) / (2000000 / (4096 * 1))
+ *             = 2.048 ms
+ *  t_WWDG_max = 1000 * ((0x7F & 0x3F) + 1) / (2000000 / (4096 * 8))
+ *             = 1048.576 ms
+ */
+
+#define ABS_DIFF_UINT(a, b)  ((a) > (b) ? (a) - (b) : (b) - (a))
+#define WWDG_TIMEOUT_ERROR_MARGIN   (100 * USEC_PER_MSEC)
+#define IS_WWDG_TIMEOUT(__TIMEOUT_GOLDEN__, __TIMEOUT__)  \
+	(ABS_DIFF_UINT(__TIMEOUT_GOLDEN__, __TIMEOUT__) < \
+	 WWDG_TIMEOUT_ERROR_MARGIN)
+
+static void wwdg_stm32_irq_config(struct device *dev);
+
+static u32_t wwdg_stm32_get_pclk(struct device *dev)
+{
+	struct device *clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	const struct wwdg_stm32_config *cfg = WWDG_STM32_CFG(dev);
+	u32_t pclk_rate;
+
+	__ASSERT_NO_MSG(clk);
+
+	clock_control_get_rate(clk, (clock_control_subsys_t *) &cfg->pclken,
+			       &pclk_rate);
+
+	return pclk_rate;
+}
+
+/**
+ * @brief Calculates the timeout in microseconds.
+ *
+ * @param dev Pointer to device structure.
+ * @param prescaler The prescaler value.
+ * @param counter The counter value.
+ * @return The timeout calculated in microseconds.
+ */
+static u32_t wwdg_stm32_get_timeout(struct device *dev, u32_t prescaler,
+				    u32_t counter)
+{
+	u32_t divider = WWDG_INTERNAL_DIVIDER * (1 << (prescaler >> 7));
+	float f_wwdg = wwdg_stm32_get_pclk(dev) / divider;
+
+	return USEC_PER_SEC * (((counter & 0x3F) + 1) / f_wwdg);
+}
+
+/**
+ * @brief Calculates prescaler & counter values.
+ *
+ * @param dev Pointer to device structure.
+ * @param timeout Timeout value in microseconds.
+ * @param prescaler Pointer to prescaler value.
+ * @param counter Pointer to counter value.
+ */
+static void wwdg_stm32_convert_timeout(struct device *dev, u32_t timeout,
+				       u32_t *prescaler,
+				       u32_t *counter)
+{
+	u32_t clock_freq = wwdg_stm32_get_pclk(dev);
+	u8_t divider = 0U;
+	u8_t shift = 3U;
+
+	/* Convert timeout to seconds. */
+	float timeout_s = (float)timeout / USEC_PER_SEC;
+	float wwdg_freq;
+
+	*prescaler = 0;
+	*counter = 0;
+
+	for (divider = 8; divider >= 1; divider >>= 1) {
+		wwdg_freq = ((float)clock_freq) / WWDG_INTERNAL_DIVIDER / divider;
+		/* +1 to ceil the result, which may lose from truncation */
+		*counter = (u32_t)(timeout_s * wwdg_freq + 1) - 1;
+		*counter |= WWDG_RESET_LIMIT;
+		*prescaler = shift << 7;
+
+		if (*counter <= WWDG_COUNTER_MAX) {
+			break;
+		}
+
+		shift--;
+	}
+}
+
+static int wwdg_stm32_setup(struct device *dev, u8_t options)
+{
+	WWDG_TypeDef *wwdg = WWDG_STM32_STRUCT(dev);
+
+	/* Deactivate running when debugger is attached. */
+	if (options & WDT_OPT_PAUSE_HALTED_BY_DBG) {
+#if defined(CONFIG_SOC_SERIES_STM32F0X)
+		LL_APB1_GRP2_EnableClock(LL_APB1_GRP2_PERIPH_DBGMCU);
+#elif defined(CONFIG_SOC_SERIES_STM32L0X)
+		LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_DBGMCU);
+#endif
```

Comment:
For some socs of the H7 series there are 2 WWDG instances with associated clocks 
WWDG1 on APB3 (Cortex-M7)
WWDG2 on APB1 (Cortex-M4)
Could be nice to support it also...

Reply:
@gon1332, if modifications for H7 are giving you headaches, you can remove for this PR. I asked support thinking this would be straight forward, since this is not the case, you can give up on this series, WWDG support will be added later. Sorry for trouble.

Reply:
The truth is that I don't have the hardware and time to study for these SOCs right now. If that's not a problem anyone could also collaborate on this PR. Otherwise, support for these SOCs could be added later.

Reply:
@gon1332, sure I understand, let add H7 later, otherwise is fine.

Reply:
seems reasonnable, ok for me